federation: Fixing runtime-config support for federation-apiserver

nikhiljindal opened the pull request:
Fixes https://github.com/kubernetes/kubernetes/issues/42587
Ref https://github.com/kubernetes/kubernetes/issues/38593

Fixing the broken `--runtime-config` flag support in federation-apiserver. Fixing the bugs and using it to disable batch and autoscaling groups. Users can enable them by passing `--runtime-config=apis/all=true` to federation-apiserver.
~This also includes a bug fix to kube-apiserver registry that allows users to disable api/v1 resources~

cc @kubernetes/sig-federation-pr-reviews

nikhiljindal committed:
Updating federation-apiserver to support runtime config and using it to disable batch and autoscaling groups

nikhiljindal committed:
Adding a test for runtime-config to federation-apiserver

nikhiljindal committed:
Autogenerated Bazel changes

nikhiljindal commented:
Adding 1.6 milestone as this is a bug fix to be included in 1.6. Let me know if anyone feels otherwise.

deads2k commented on pkg/registry/batch/rest/storage_batch.go:
```diff
@@ -52,15 +54,21 @@ func (p RESTStorageProvider) NewRESTStorage(apiResourceConfigSource serverstorag
+}
+
+// Adds rest storage interfaces for v1/jobs resource to the given storage map.
+func (p RESTStorageProvider) AddV1JobsStorage(apiResourceConfigSource serverstorage.APIResourceConfigSource, restOptionsGetter generic.RESTOptionsGetter, storage map[string]rest.Storage) {
```
Don't mutate your args.  It makes for really confusing code.

deads2k replied:
This function shouldn't be making a decision about enablement.  You asked to add them, you should add them.

nikhiljindal replied:
I want a method that also has the check. Will renaming this to AddV1JobsStorageIfEnabled help?

nikhiljindal replied:
Updated to return the map

nikhiljindal replied:
reverted changes to registry

deads2k commented on pkg/registry/core/rest/storage_core.go:
```diff
@@ -179,50 +175,90 @@ func (c LegacyRESTStorageProvider) NewLegacyRESTStorage(restOptionsGetter generi
 
 	serviceRest := service.NewStorage(serviceRegistry, endpointRegistry, ServiceClusterIPAllocator, ServiceNodePortAllocator, c.ProxyTransport)
 
-	restStorageMap := map[string]rest.Storage{
```
This change is really significant.  Disabling bits of the core API isn't a thing that was generally support before.  I remember the code only allowed it for extensions and even that was probably a mistake.

I don't think that you should be able to pick and choose from groups as a rule.

deads2k replied:
The option may have parsed before, but this group has never actually respected individual enablement.  I don't think we should start now.

nikhiljindal replied:
Why not? Core is too broad. I want to disable just ReplicationControllers for example.
runtime-config allows users to disable groups, group versions or specific resources.
It will be confusing if some groups allow disabling specific resources while some groups do not.
I think this group not allowing disabling specific resources was actually a bug.

nikhiljindal replied:
On more research and help from @liggitt What you are saying is correct: https://github.com/kubernetes/kubernetes/pull/42303#issuecomment-283230144. Will update the PR

nikhiljindal replied:
Updated the PR to revert this change

sttts commented on pkg/registry/core/rest/storage_core.go:
```diff
@@ -179,50 +175,90 @@ func (c LegacyRESTStorageProvider) NewLegacyRESTStorage(restOptionsGetter generi
-		"configMaps":                    configMapStorage,
-
-		"componentStatuses": componentstatus.NewStorage(componentStatusStorage{c.StorageFactory}.serversToValidate),
+	restStorageMap := map[string]rest.Storage{}
```
If we need this logic, isn't there some generic way to express it other than turning this func into something unreadable?

nikhiljindal replied:
reverted changes in this file.

sttts commented on pkg/registry/batch/rest/storage_batch.go:
```diff
@@ -52,15 +54,21 @@ func (p RESTStorageProvider) NewRESTStorage(apiResourceConfigSource serverstorag
 		storage["jobs"] = jobsStorage
 		storage["jobs/status"] = jobsStatusStorage
+	} else {
+		glog.V(1).Infof("Skipping disabled resource batch/v1/jobs")
```
There should be some generic way to output these message, not one Infof for each resource.

nikhiljindal replied:
Moved it to a common function

sttts commented on federation/cmd/federation-apiserver/app/federation.go:
```diff
@@ -22,20 +22,33 @@ import (
+func installFederationAPIs(g *genericapiserver.GenericAPIServer, optsGetter generic.RESTOptionsGetter, apiResourceConfigSource storage.APIResourceConfigSource) {
+	groupName := federation.GroupName
+	if !apiResourceConfigSource.AnyResourcesForGroupEnabled(groupName) {
+		glog.V(1).Infof("Skipping disabled API group %q", groupName)
```
This also looks like a lot of boilerplate code that should be done in a generic way somehow. The previously completely declarative style with one InstallAPIGroup style turned into some complex logic for each group. This will drift apart pretty fast.

nikhiljindal replied:
Agreed. Updated to move the boilerplate code into a common method.

liggitt commented:
cc @kubernetes/sig-api-machinery-pr-reviews

ethernetdan commented:
Removing PR from 1.6 milestone since code freeze has begun:
* If this is a bug fix, please label as `kind/bug` and reapply the milestone.
* If this is a feature for 1.6, an [exception](https://github.com/kubernetes/features/blob/master/EXCEPTIONS.md) should be filed for it before 11:00PM PST on 3/2. Note that most exceptions to code freeze are not accepted.
* If this is a feature for 1.7, feel free to add the 1.7 milestone to the PR.

/cc @kubernetes/release-team @kubernetes/kubernetes-release-managers

calebamiles commented:
I would really like to move this PR into the next milestone @nikhiljindal. It did not have an `lgtm` by code freeze and there are existing questions to be addressed from the review. There is also no additional test coverage in the PR nor any mention of the specific tests that the PR is supposed to address as a bug fix. Please add this information when available. Thanks!

**cc: @ethernetdan, @kubernetes/kubernetes-release-managers**

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/42225)
<!-- Reviewable:end -->

nikhiljindal commented:
Thanks for the great comments @deads2k and @sttts 
Updated the code as per comments.

Have also reverted changes to registry or other kubernetes code. The changes now are only limited to federation. PTAL

sttts commented on federation/cmd/federation-apiserver/app/autoscaling.go:
```diff
@@ -21,24 +21,35 @@ import (
-		"horizontalpodautoscalers/status": hpaStatusStorage,
+func installAutoscalingAPIs(g *genericapiserver.GenericAPIServer, optsGetter generic.RESTOptionsGetter, apiResourceConfigSource storage.APIResourceConfigSource) {
+	groupName := autoscaling.GroupName
+	hpaStorageFn := func() map[string]rest.Storage {
```
why a func? Why not just inline?

nikhiljindal replied:
Yes it does need to be a func, but I wrote it that way because I had some future refactoring in mind where I want this to be shared between kubernetes and federation. A single function that returns the storage for a resource (for ex storages for hpa and its subresources in this case).
I didnt do that refactoring in this PR but I do have plans for a follow up PR.

So if you think this is bad, then I can change this to inline for now and will change them to funcs later.
But if its fine then I will like to keep them as is. Let me know.

sttts replied:
It's not bad, only looks like there is some reason like side-effects behind that is not obvious.

sttts commented on federation/cmd/federation-apiserver/app/autoscaling.go:
```diff
@@ -21,24 +21,35 @@ import (
-		"horizontalpodautoscalers":        hpaStorage,
-		"horizontalpodautoscalers/status": hpaStatusStorage,
+func installAutoscalingAPIs(g *genericapiserver.GenericAPIServer, optsGetter generic.RESTOptionsGetter, apiResourceConfigSource storage.APIResourceConfigSource) {
+	groupName := autoscaling.GroupName
```
Unnecessary variable. Only used once. Inline.

nikhiljindal replied:
removed

sttts commented on federation/cmd/federation-apiserver/app/core.go:
```diff
@@ -41,28 +42,56 @@ import (
+		}
+	}
+	eventsStorageFn := func() map[string]rest.Storage {
+		eventStore := eventstore.NewREST(optsGetter, uint64(s.EventTTL.Seconds()))
```
are these having any side effects? Or why are they a func and not just vars?

nikhiljindal replied:
no side effect. same as above

sttts commented on federation/cmd/federation-apiserver/app/install.go:
```diff
@@ -0,0 +1,55 @@
+// Filters the resources from the given resources storage map to those that are enabled in the given apiResourceConfigSource.
+// resourcesStorageMap is expected to contain all resources in a group version.
+// Returns false if none of the resources are enabled and hence the whole group version should be disabled.
+func getEnabledResources(groupVersion schema.GroupVersion, resourcesStorageMap map[string]getResourcesStorageFunc, apiResourceConfigSource storage.APIResourceConfigSource) (bool, map[string]rest.Storage) {
```
s/getE/e/. `get*` is a Go anti-pattern.

nikhiljindal replied:
hmm hadnt realized that. Fixed. Any link to docs that say that this is an anti-pattern?

sttts replied:
https://golang.org/doc/effective_go.html#Getters

sttts commented on federation/cmd/federation-apiserver/app/server.go:
```diff
@@ -108,8 +111,7 @@ func Run(s *options.ServerRunOptions) error {
 
-	// TODO: register cluster federation resources here.
-	resourceConfig := serverstorage.NewResourceConfig()
+	resourceConfig := getDefaultResourceConfig()
```
s/getD/d/

nikhiljindal replied:
fixed

sttts commented on test/integration/federation/server_test.go:
```diff
@@ -69,11 +76,32 @@ func TestRun(t *testing.T) {
+
+// Verifies that all APIs are enabled when runtime config is set to all.
+func TestRunWithRuntimeConfigAll(t *testing.T) {
+	securePort = 6443 + 5
```
Use framework.LocalPort from https://github.com/kubernetes/kubernetes/pull/42278/files#diff-efb71138113f6fe331f73e11c05e323aR519

nikhiljindal replied:
I will let that PR get merged and update this in a follow up PR? We have an open issue for that https://github.com/kubernetes/kubernetes/issues/42400

sttts replied:
If this is for 1.7, there is no hurry to get it merged. I would prefer to have this fixed to avoid flakes.

nikhiljindal replied:
I would like to get it in 1.6 if the release team agrees.
To clarify, this should not lead to flakes. Either the tests will always fail or pass.

sttts replied:
If there is a follow-up soon'ish for master, I can live with it if it has to get into 1.6.

sttts replied:
Or just cherry-pick https://github.com/kubernetes/kubernetes/pull/42278/commits/e51262888227119d1c2b1fe8c64cec209e254150 into your PR here and this is ready to go. Github will probably figure out the repeated commit itself.

sttts commented on test/integration/federation/server_test.go:
```diff
@@ -256,7 +298,7 @@ func testCoreResourceList(t *testing.T) {
 	assert.Equal(t, "", apiResourceList.APIVersion)
 	assert.Equal(t, v1.SchemeGroupVersion.String(), apiResourceList.GroupVersion)
-	// Assert that there are exactly 7 resources.
+	// Assert that there are exactly 8 resources.
```
drop the comment. No additional info.

nikhiljindal replied:
removed

sttts commented:
@nikhiljindal left another couple of comments. But it looks much better already 👍

nikhiljindal commented:
Thanks for the review. Updated code as per comments. PTAL

marun commented:
I think #42278 should merge before this.

nikhiljindal commented:
@marun I want to see #42278 get merged soon as well. But I dont think we should block bug fixes for it.

sttts commented on federation/cmd/federation-apiserver/app/install.go:
```diff
@@ -0,0 +1,55 @@
+
+// Filters the resources from the given resources storage map to those that are enabled in the given apiResourceConfigSource.
+// resourcesStorageMap is expected to contain all resources in a group version.
+// Returns false if none of the resources are enabled and hence the whole group version should be disabled.
```
the bool is equivalent to `len(enabledResources) == 0` ? I would just remove it.

nikhiljindal replied:
Isnt it better to keep that logic in this function rather than requiring all callers to do that?

sttts replied:
You have a point, this was meant to keep the other code clean. I am fine with it.

sttts commented:
@nikhiljindal only a nit left about `len(enabledResources) == 0` and about the `localPort`. After that lgtm.

nikhiljindal commented:
Now that https://github.com/kubernetes/kubernetes/pull/42664 is in, this can wait till after 1.6

nikhiljindal commented:
Rebased and updated to use the test fixture from https://github.com/kubernetes/kubernetes/pull/42278.
This should be ready to go now.

sttts commented:
/lgtm
